src/lib.rs

```diff
@@ -439,114 +387,92 @@ impl<M: Manager> Pool<M> {
-        let mut c = self.get_or_create_conn().await?;
-        c.last_used_at = Instant::now();
+        let _guard = GaugeGuard::increment(WAIT_COUNT);
+        let c = self.get_or_create_conn().await?;
 
         let conn = Connection {
-            pool: Some(self.clone()),
+            pool: self.clone(),
             conn: Some(c),
         };
 
-        gauge!(ACTIVE_CONNECTIONS).increment(1.0);
-        gauge!(WAIT_COUNT).decrement(1.0);
-
         Ok(conn)
     }
 
     async fn validate_conn(
         &self,
         internal_config: InternalConfig,
-        conn: &mut Conn<M::Connection, M::Error>,
-    ) -> bool {
-        if conn.brand_new {
-            return true;
+        conn: IdleConn<M::Connection>,
+    ) -> Option<IdleConn<M::Connection>> {
+        if conn.is_brand_new() {
+            return Some(conn);
         }
 
         if conn.expired(internal_config.max_lifetime) {
-            return false;
+            return None;
         }
 
         if conn.idle_expired(internal_config.max_idle_lifetime) {
-            return false;
+            return None;
         }
 
         let needs_health_check = self.0.config.health_check
             && conn.needs_health_check(self.0.config.health_check_interval);
 
         if needs_health_check {
-            let raw = conn.raw.take().unwrap();
-            match self.0.manager.check(raw).await {
-                Ok(raw) => {
-                    conn.last_checked_at = Instant::now();
-                    conn.raw = Some(raw)
-                }
-                Err(_e) => return false,
-            }
+            let (raw, split) = conn.split_raw();
+            let checked_raw = self.0.manager.check(raw).await.ok()?;
+            let mut checked = split.restore(checked_raw);
+            checked.mark_checked();
+            return Some(checked);
         }
-        true
+        Some(conn)
     }
 
-    async fn get_or_create_conn(&self) -> Result<Conn<M::Connection, M::Error>, Error<M::Error>> {
+    async fn get_or_create_conn(&self) -> Result<ActiveConn<M::Connection>, Error<M::Error>> {
         self.0.state.wait_count.fetch_add(1, Ordering::Relaxed);
-        gauge!(WAIT_COUNT).increment(1.0);
-        let wait_start = Instant::now();
+        let wait_guard = DurationHistogramGuard::start(WAIT_DURATION);
 
-        let permit = self
-            .0
-            .semaphore
-            .acquire()
+        let semaphore = Arc::clone(&self.0.semaphore);
+        let permit = semaphore
+            .acquire_owned()
             .await
             .map_err(|_| Error::PoolClosed)?;
 
         self.0.state.wait_count.fetch_sub(1, Ordering::SeqCst);
 
         let mut internals = self.0.internals.lock().await;
 
-        internals.wait_duration += wait_start.elapsed();
-        histogram!(WAIT_DURATION).record(wait_start.elapsed());
+        internals.wait_duration += wait_guard.elapsed();
+        drop(wait_guard);
```

Comment:
A bit of an edge case and I'm not sure how critical it is here, but if the thread is preempted by the kernel between these two statements, the counters in `internal` and the histogram may diverge a bit. Could be sensitive to running on a cloud VM with something like 100m CPU quota.

Reply:
yep `internals.wait_duration += wait_guard.into_elapsed()` sounds great